src/sentry/debug_files/artifact_bundles.py

```diff
@@ -61,176 +58,84 @@ def index_artifact_bundles_for_release(
-    # irrespectively if we have urls to add for that bundle.
-    for artifact_bundle in artifact_bundles:
-        urls = urls_by_bundle.get(artifact_bundle, set())
-
         try:
-            # We want to start a transaction for each bundle, so that in case of failures we keep consistency at the
-            # bundle level, and we also have to retry only the failed bundle in the future and not all the bundles.
-            with atomic_transaction(
-                using=(
-                    router.db_for_write(ArtifactBundle),
-                    router.db_for_write(ArtifactBundleIndex),
-                )
+            if not set_artifact_bundle_being_indexed_if_null(
+                organization_id=organization_id, artifact_bundle_id=artifact_bundle.id
             ):
-                # Since we use read committed isolation, the value we read here can change after query execution, but
-                # we have this check in place for analytics purposes.
-                is_bundle_not_indexed = ArtifactBundle.objects.filter(
-                    id=artifact_bundle.id,
-                    indexing_state=ArtifactBundleIndexingState.NOT_INDEXED.value,
-                ).exists()
-                # If the bundle was already indexed, we will skip insertion into the database.
-                if not is_bundle_not_indexed:
-                    metrics.incr("artifact_bundle_indexing.bundle_was_already_indexed")
-                    continue
-
-                # Now we index the urls in the bundle.
-                _index_urls_in_bundle(
-                    organization_id=organization_id,
-                    artifact_bundle=artifact_bundle,
-                    release=release,
-                    dist=dist,
-                    date_added=date_added,
-                    urls=urls,
-                )
+                # A different asynchronous job is taking care of this bundle.
+                metrics.incr("artifact_bundle_indexing.bundle_already_being_indexed")
+                continue
 
-            # After the transaction was successful we could clean the redis cache, but it's fine to keep the value in
-            # there and wait for auto-expiration.
-            metrics.incr("artifact_bundle_indexing.bundles_indexed")
-            metrics.incr("artifact_bundle_indexing.urls_indexed", len(urls))
+            _index_urls_in_bundle(organization_id, artifact_bundle, release, dist)
         except Exception as e:
             # We want to catch the error and continue execution, since we can try to index the other bundles.
             metrics.incr("artifact_bundle_indexing.index_single_artifact_bundle_error")
             sentry_sdk.capture_exception(e)
 
+            # TODO: Do we want to `remove_artifact_bundle_indexing_state` here so that
+            # a different job can retry this? Probably not, as we want to at the very least
+            # debounce this in case there is a persistent error?
```

Comment:
Exactly, this was the reasoning behind not deleting it. We would like to give the system some head room to avoid failures in a fast sequence, however, we should track the errors and in case we see that they are quite frequent, develop some sort of circuit breaker mechanism after some retries.